cirq/sim/clifford/clifford_simulator.py

```diff
@@ -300,25 +316,50 @@ def apply_unitary(self, op: 'cirq.Operation'):
                              self.qubit_map[op.qubits[1]])
             self.ch_form._CZ(self.qubit_map[op.qubits[0]],
                              self.qubit_map[op.qubits[1]])
-        elif op.gate == cirq.Z:
-            self.tableau._Z(self.qubit_map[op.qubits[0]])
-            self.ch_form._Z(self.qubit_map[op.qubits[0]])
-        elif op.gate == cirq.X:
-            self.tableau._X(self.qubit_map[op.qubits[0]])
-            self.ch_form._X(self.qubit_map[op.qubits[0]])
-        elif op.gate == cirq.Y:
-            self.tableau._Y(self.qubit_map[op.qubits[0]])
-            self.ch_form._Y(self.qubit_map[op.qubits[0]])
-        elif op.gate == cirq.S:
-            self.tableau._S(self.qubit_map[op.qubits[0]])
-            self.ch_form._S(self.qubit_map[op.qubits[0]])
-        elif op.gate == cirq.H:
-            self.tableau._H(self.qubit_map[op.qubits[0]])
-            self.ch_form._H(self.qubit_map[op.qubits[0]])
         else:
-            raise ValueError('%s cannot be run with Clifford simulator' %
+            raise ValueError('%s cannot be run with Clifford simulator.' %
                              str(op.gate))  # type: ignore
 
+    def apply_single_qubit_unitary(self, op: 'cirq.Operation'):
+        assert op.gate is not None
+        u = unitary(op.gate)
+        clifford_gate = SingleQubitCliffordGate.from_unitary(u)
+        if clifford_gate is None:
+            raise ValueError('%s cannot be run with Clifford simulator.' %
+                             str(op.gate))
+
+        h = unitary(ops.H)
+        s = unitary(ops.S)
+        applied_unitary = np.eye(2)
+        for axis, quarter_turns in clifford_gate.decompose_rotation():
+            for _ in range(quarter_turns % 4):
+                if axis == pauli_gates.X:
+                    self._apply_H(self.qubit_map[op.qubits[0]])
+                    self._apply_S(self.qubit_map[op.qubits[0]])
+                    self._apply_H(self.qubit_map[op.qubits[0]])
+                    applied_unitary = h @ s @ h @ applied_unitary
+                elif axis == pauli_gates.Y:
+                    self._apply_S(self.qubit_map[op.qubits[0]])
+                    self._apply_S(self.qubit_map[op.qubits[0]])
+                    self._apply_H(self.qubit_map[op.qubits[0]])
+                    applied_unitary = h @ s @ s @ applied_unitary
+                else:
+                    assert axis == pauli_gates.Z
+                    self._apply_S(self.qubit_map[op.qubits[0]])
```

Comment:
I like this implementation much better.

A small question, would it be too messy to check for the gate to be `H` beforehand? I find it unpleasant that an H gate, even though very natively supported, will get resolved to something like `hshhshhss`.

Reply:
Added special handling for H gate.